cmd/fleet/main.go

```diff
@@ -184,3 +160,142 @@ func NewCommand(version string) *cobra.Command {
 	cmd.Flags().StringP("config", "c", "fleet-server.yml", "Configuration for Fleet Server")
 	return cmd
 }
+
+type fleetServer struct {
+	cfg *config.Config
+
+	es *es.Client
+	sv saved.CRUD
+
+	lock           sync.Mutex
+	profilerCancel context.CancelFunc
+	serverCancel   context.CancelFunc
+	errCh          chan error
+}
+
+// NewFleetServer creates the actual fleet server service.
+func NewFleetServer(ctx context.Context, cfg *config.Config) (*fleetServer, error) {
+	es, err := es.New(ctx, cfg)
+	if err != nil {
+		return nil, err
+	}
+	sv := saved.NewMgr(es.Bulk(), savedObjectKey())
+	return &fleetServer{
+		cfg: cfg,
+		es:  es,
+		sv:  sv,
+	}, nil
+}
+
+// Run runs the fleet server.
+func (f *fleetServer) Run(ctx context.Context) error {
+	// Initial indices bootstrapping, needed for agents actions development
+	// TODO: remove this after the indices bootstrapping logic implemented in ES plugin
+	err := esboot.EnsureESIndices(ctx, f.es)
+	if err != nil {
+		return err
+	}
+
+	// profiler restarts on reload
+	profilerCtx, profilerCancel := context.WithCancel(ctx)
+	err = profile.RunProfiler(profilerCtx, f.cfg.Inputs[0].Server.Profile.Bind)
+	if err != nil {
+		profilerCancel()
+		return err
+	}
+
+	// server restarts on reload
+	errCh := make(chan error)
+	serverCancel, err := f.runServer(ctx, errCh)
+	if err != nil {
+		profilerCancel()
+		serverCancel()
+		return err
+	}
+	f.lock.Lock()
+	f.profilerCancel = profilerCancel
+	f.serverCancel = serverCancel
```

Comment:
maybe could use one cancellable context for both server and profiler?

Reply:
then it would restart both, when either changed

you might want to turn on profiling while the server is already running, but if you restart the server all the current status of the server will be lost and the information you are trying to get by turning on the profiler might be lost